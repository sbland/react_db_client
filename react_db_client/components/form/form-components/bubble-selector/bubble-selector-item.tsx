import React from 'react';
import PropTypes from 'prop-types';
import { BubbleSelectorListItemStyle } from './styles';

export interface IBubbleSelectorItemProps {
  uid: string | number;
  label: string | React.ReactNode;
  handleSelect: (uid: string | number) => void;
  isSelected: boolean;
}

export const BubbleSelectorItem = ({
  uid,
  label,
  handleSelect,
  isSelected,
}: IBubbleSelectorItemProps) => (
  <BubbleSelectorListItemStyle className="bubbleSelector_item">
    <button
      type="button"
      className={isSelected ? 'button-two selected' : 'button-one notSelected'}
      onClick={() => handleSelect(uid)}
      data-testid={`bubbleSelector-item-${isSelected ? "selected" : "unselected"}`}
    >
      {label}
    </button>
  </BubbleSelectorListItemStyle>
);

BubbleSelectorItem.propTypes = {
  uid: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  label: PropTypes.string.isRequired,
  handleSelect: PropTypes.func.isRequired,
  isSelected: PropTypes.bool,
};

BubbleSelectorItem.defaultProps = {
  isSelected: false,
};
