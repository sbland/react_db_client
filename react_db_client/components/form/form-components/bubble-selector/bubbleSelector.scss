$unit: 1rem;//var(--unit, 1rem);

.bubbleSelector_list {
  list-style: none;
  display: flex;
  flex-wrap: wrap;

  li {
    margin-right: ($unit * 0.5);
    margin-bottom: ($unit * 0.1);
  }

  &.unselected {
    max-height: ($unit * 8);
    overflow-y: auto;
  }
}

.bubbleSelector_item {
  button {
    border-radius: ($unit * 0.2);
    min-width: ($unit * 4);
  }
}
