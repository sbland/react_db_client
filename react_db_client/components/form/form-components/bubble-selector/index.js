export { BubbleSelector } from './bubble-selector';
