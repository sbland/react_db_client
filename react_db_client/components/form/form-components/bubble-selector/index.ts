export { BubbleSelector } from './bubble-selector';
export type { EActions, IOpt, IBubbleSelectorProps } from './bubble-selector';
