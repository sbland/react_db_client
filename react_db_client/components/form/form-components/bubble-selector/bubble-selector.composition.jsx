import React, { useState } from 'react';
import { BubbleSelector } from './bubble-selector';
import { demoItems } from './demo-data';

export const BasicBubbleSelector = () => {
  const [activeSelection, setActiveSelection] = useState(['a', 'b']);
  return (
    <BubbleSelector
      activeSelection={activeSelection}
      options={demoItems}
      updateActiveSelection={setActiveSelection}
    />
  );
};


export const BasicBubbleSelectorCustomElement = () => {
  const [activeSelection, setActiveSelection] = useState(['a', 'b', 'e']);
  return (
    <BubbleSelector
      activeSelection={activeSelection}
      options={demoItems}
      updateActiveSelection={setActiveSelection}
    />
  );
};

export const BasicBubbleSelectorGrouped = () => {
  const [activeSelection, setActiveSelection] = useState(['a', 'b']);
  return (
    <BubbleSelector
      activeSelection={activeSelection}
      options={demoItems}
      updateActiveSelection={setActiveSelection}
      groupSelected
      // hideUnselected
    />
  );
};

export const BubbleSelectorManualInput = () => {
  const [activeSelection, setActiveSelection] = useState(['a', 'b', 'e']);
  return (
    <BubbleSelector
      activeSelection={activeSelection}
      options={demoItems}
      updateActiveSelection={setActiveSelection}
      allowManualInput
    />
  );
};
