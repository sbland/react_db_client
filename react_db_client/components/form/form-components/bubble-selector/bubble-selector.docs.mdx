---
description: 'A BubbleSelector component.'
labels: ['field-component', 'multi-select', 'form']
---

import { BubbleSelector } from './bubble-selector';
import { BasicBubbleSelector } from './bubble-selector.composition';

## React Component for rendering selectable list items

Bubble selector for selecting multiple items in a list of options.

### Using props to customize the text

Modify the text to see it change live:
```js live
<BasicBubbleSelector />
```
