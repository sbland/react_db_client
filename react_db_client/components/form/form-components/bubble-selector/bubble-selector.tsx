import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { BubbleSelectorItem } from './bubble-selector-item';
import { BubbleSelectorListStyle } from './styles';

export enum EActions {
  REMOVE = 'remove',
  ADD = 'add',
}

export interface IOpt {
  uid: string | number;
  label: string | React.ReactNode;
}

export interface IBubbleSelectorProps {
  activeSelection: (string | number)[];
  updateActiveSelection: (
    newItems: (string | number)[],
    action?: EActions,
    uid?: string | number
  ) => void;
  options: IOpt[];
  isSorted?: boolean;
  groupSelected?: boolean;
  allowManualInput?: boolean;
  allowSelectAll?: boolean;
  hideUnselected?: boolean;
}

export const BubbleSelector = ({
  activeSelection,
  updateActiveSelection,
  options,
  isSorted,
  groupSelected,
  allowManualInput,
  allowSelectAll,
  hideUnselected,
}: IBubbleSelectorProps) => {
  const [showUnselected, setShowUnselected] = useState(!hideUnselected);
  const [manualInput, setManualInput] = useState('');

  const handleSelect = (uid) => {
    const indexInSelection = activeSelection.indexOf(uid);
    const copyOfActiveSelection = [...activeSelection];
    let action: EActions | null = null;
    if (indexInSelection !== -1) {
      action = EActions.REMOVE;
      copyOfActiveSelection.splice(indexInSelection, 1);
    } else {
      action = EActions.ADD;
      copyOfActiveSelection.push(uid);
    }
    const sortedList = options
      .map((opt) => opt.uid)
      .filter((opt) => copyOfActiveSelection.indexOf(opt) !== -1);
    updateActiveSelection(sortedList, action, uid);
  };

  const handleSelectAll = () => {
    updateActiveSelection(options.map((o) => o.uid));
  };

  const selectedItems = activeSelection.map(
    (sel) => options.filter((opt) => opt.uid === sel)[0] || { uid: sel, label: sel }
  );
  const unselectedItems = options.filter((opt) => activeSelection.indexOf(opt.uid) === -1);

  const additionalItems = selectedItems.filter(
    (opt) => options.map((o) => o.uid).indexOf(opt.uid) === -1
  );

  const allItems = groupSelected
    ? selectedItems.concat(unselectedItems)
    : [...options, ...additionalItems];

  const sortFn = isSorted
    ? (a: IOpt, b: IOpt) =>
        String(a?.label).toLowerCase() > String(b.label).toLowerCase() ? 1 : -1
    : () => -1;

  const mapAllItems = allItems
    .sort(sortFn)
    .map((opt) => (
      <BubbleSelectorItem
        key={opt.uid}
        uid={opt.uid}
        label={opt.label}
        handleSelect={handleSelect}
        isSelected={activeSelection.indexOf(opt.uid) !== -1}
      />
    ));

  const mapSelectedItems = selectedItems
    .sort(sortFn)
    .map((opt) => (
      <BubbleSelectorItem
        key={opt.uid}
        uid={opt.uid}
        label={opt.label}
        handleSelect={handleSelect}
        isSelected
      />
    ));

  const mapUnselectedItems = unselectedItems
    .sort(sortFn)
    .map((opt) => (
      <BubbleSelectorItem
        key={opt.uid}
        uid={opt.uid}
        label={opt.label}
        handleSelect={handleSelect}
      />
    ));

  return (
    <div className="bubbleSelector">
      {allowManualInput && (
        <input
          type="text"
          value={manualInput || ''}
          onChange={(e) => setManualInput(e.target.value)}
          onKeyPress={(e) => {
            if (e.key === 'Enter') {
              handleSelect(manualInput);
              setManualInput('');
            }
          }}
        />
      )}
      {groupSelected && (
        <>
          <BubbleSelectorListStyle
            className="bubbleSelector_list selected"
            data-testid="bubbleSelector_list-selected"
          >
            {mapSelectedItems}
          </BubbleSelectorListStyle>
          {showUnselected && (
            <>
              <BubbleSelectorListStyle
                className="bubbleSelector_list unselected"
                data-testid="bubbleSelector_list-notselected"
              >
                {mapUnselectedItems}
              </BubbleSelectorListStyle>
              {hideUnselected && (
                <button
                  type="button"
                  className="button-one"
                  onClick={() => setShowUnselected(false)}
                >
                  -
                </button>
              )}
            </>
          )}
          {!showUnselected && (
            <p>
              <button type="button" className="button-one" onClick={() => setShowUnselected(true)}>
                + more items
              </button>
            </p>
          )}
        </>
      )}
      {!groupSelected && (
        <BubbleSelectorListStyle className="bubbleSelector_list notGrouped" data-testid="bubbleSelector_list">
          {mapAllItems}
        </BubbleSelectorListStyle>
      )}
      {allowSelectAll && (
        <button type="button" className="button-two" onClick={handleSelectAll}>
          Select All
        </button>
      )}
    </div>
  );
};

BubbleSelector.propTypes = {
  activeSelection: PropTypes.arrayOf(PropTypes.string),
  options: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
    })
  ).isRequired,
  updateActiveSelection: PropTypes.func.isRequired,
  isSorted: PropTypes.bool,
  groupSelected: PropTypes.bool,
  allowManualInput: PropTypes.bool,
  allowSelectAll: PropTypes.bool,
  hideUnselected: PropTypes.bool,
};

BubbleSelector.defaultProps = {
  activeSelection: [],
  isSorted: false,
  groupSelected: false,
  allowManualInput: false,
  allowSelectAll: false,
  hideUnselected: false,
};
