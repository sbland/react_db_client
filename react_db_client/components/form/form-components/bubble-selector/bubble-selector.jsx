import React, { useState } from 'react';
import PropTypes from 'prop-types';

import './bubbleSelector.scss';

export const BubbleSelectorItem = ({ uid, label, handleSelect, isSelected }) => (
  <li className="bubbleSelector_item">
    <button
      type="button"
      className={isSelected ? 'button-two selected' : 'button-one notSelected'}
      onClick={() => handleSelect(uid)}
    >
      {label}
    </button>
  </li>
);

BubbleSelectorItem.propTypes = {
  uid: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  handleSelect: PropTypes.func.isRequired,
  isSelected: PropTypes.bool,
};

BubbleSelectorItem.defaultProps = {
  isSelected: false,
};

export const BubbleSelector = ({
  activeSelection,
  updateActiveSelection,
  options,
  isSorted,
  groupSelected,
  allowManualInput,
  allowSelectAll,
}) => {
  const [showUnselected, setShowUnselected] = useState(false);
  const [manualInput, setManualInput] = useState('');

  const handleSelect = (uid) => {
    const indexInSelection = activeSelection.indexOf(uid);
    const copyOfActiveSelection = [...activeSelection];
    let action = null;
    if (indexInSelection !== -1) {
      action = 'remove';
      copyOfActiveSelection.splice(indexInSelection, 1);
    } else {
      action = 'add';
      copyOfActiveSelection.push(uid);
    }
    const sortedList = options
      .map((opt) => opt.uid)
      .filter((opt) => copyOfActiveSelection.indexOf(opt) !== -1);
    updateActiveSelection(sortedList, action, uid);
  };

  const handleSelectAll = () => {
    updateActiveSelection(options.map((o) => o.uid))
  }

  const selectedItems = activeSelection.map(
    (sel) => options.filter((opt) => opt.uid === sel)[0] || { uid: sel, label: sel }
  );
  // console.log(selectedItems);
  // const selectedItems = options.filter((opt) => (activeSelection.indexOf(opt.uid) !== -1));
  const unselectedItems = options.filter((opt) => activeSelection.indexOf(opt.uid) === -1);

  const allItems = selectedItems.concat(unselectedItems);

  const sortFn = isSorted ? (a, b) => a.label.toUpperCase() > b.label.toUpperCase() : () => false;

  const mapAllItems = allItems
    .sort(sortFn)
    .map((opt) => (
      <BubbleSelectorItem
        key={opt.uid}
        uid={opt.uid}
        label={opt.label}
        handleSelect={handleSelect}
        isSelected={activeSelection.indexOf(opt.uid) !== -1}
      />
    ));

  const mapSelectedItems = selectedItems
    .sort(sortFn)
    .map((opt) => (
      <BubbleSelectorItem
        key={opt.uid}
        uid={opt.uid}
        label={opt.label}
        handleSelect={handleSelect}
        isSelected
      />
    ));

  const mapUnselectedItems = unselectedItems
    .sort(sortFn)
    .map((opt) => (
      <BubbleSelectorItem
        key={opt.uid}
        uid={opt.uid}
        label={opt.label}
        handleSelect={handleSelect}
      />
    ));

  return (
    <div className="bubbleSelector">
      {allowManualInput && (
        <input
          type="text"
          value={manualInput || ''}
          onChange={(e) => setManualInput(e.target.value)}
          onKeyPress={(e) => {
            if (e.key === 'Enter') {
              handleSelect(manualInput);
              setManualInput('');
            }
          }}
        />
      )}
      {groupSelected && (
        <>
          <ul className="bubbleSelector_list selected">{mapSelectedItems}</ul>
          {showUnselected && (
            <>
              <ul className="bubbleSelector_list unselected">{mapUnselectedItems}</ul>
              <button type="button" className="button-one" onClick={() => setShowUnselected(false)}>
                -
              </button>
            </>
          )}
          {!showUnselected && (
            <p>
              <button type="button" className="button-one" onClick={() => setShowUnselected(true)}>
                + more items
              </button>
            </p>
          )}
        </>
      )}
      {!groupSelected && <ul className="bubbleSelector_list">{mapAllItems}</ul>}
      {allowSelectAll && (
        <button type="button" className="button-two" onClick={handleSelectAll}>
          Select All
        </button>
      )}
    </div>
  );
};

BubbleSelector.propTypes = {
  activeSelection: PropTypes.arrayOf(PropTypes.string),
  options: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
    })
  ).isRequired,
  updateActiveSelection: PropTypes.func.isRequired,
  isSorted: PropTypes.bool,
  groupSelected: PropTypes.bool,
  allowManualInput: PropTypes.bool,
  allowSelectAll: PropTypes.bool,
};

BubbleSelector.defaultProps = {
  activeSelection: [],
  isSorted: false,
  groupSelected: false,
  allowManualInput: false,
  allowSelectAll: false,
};
