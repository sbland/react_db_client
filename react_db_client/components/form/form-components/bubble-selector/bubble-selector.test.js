import '@samnbuk/react_db_client.helpers.enzyme-setup';
import React from 'react';
import { shallow, mount } from 'enzyme';
import { BubbleSelector } from './bubble-selector';

describe('BubbleSelector', () => {
  it('Renders', () => {
    shallow(
      <BubbleSelector
        activeSelection={['a', 'b']}
        options={[
          { uid: 'a', label: 'a' },
          { uid: 'b', label: 'b' },
          { uid: 'c', label: 'c' },
          { uid: 'd', label: 'd' },
        ]}
        updateActiveSelection={() => {}}
      />
    );
  });
  it('Matches Snapshot', () => {
    const component = shallow(
      <BubbleSelector
        activeSelection={['a', 'b']}
        options={[
          { uid: 'a', label: 'a' },
          { uid: 'b', label: 'b' },
          { uid: 'c', label: 'c' },
          { uid: 'd', label: 'd' },
        ]}
        updateActiveSelection={() => {}}
      />
    );
    const tree = component.debug();
    expect(tree).toMatchSnapshot();
  });
  describe('selecting', () => {
    const updateActiveSelection = jest.fn();
    const bubbleSelector = mount(
      <BubbleSelector
        activeSelection={['a', 'b']}
        options={[
          { uid: 'a', label: 'a' },
          { uid: 'b', label: 'b' },
          { uid: 'c', label: 'c' },
          { uid: 'd', label: 'd' },
        ]}
        updateActiveSelection={updateActiveSelection}
      />
    );
    it('highlights selected', () => {
      expect(bubbleSelector.find('.selected').length).toEqual(2);
    });
    it('calls update selection with item removed when selected item is clicked', () => {
      const selectedItemBtn = bubbleSelector.find('.selected').first();
      selectedItemBtn.simulate('click');
      expect(updateActiveSelection).toHaveBeenCalled();
      expect(updateActiveSelection).toHaveBeenCalledWith(['b'], 'remove', 'a');
      updateActiveSelection.mockClear();
    });
    it('calls update selection with item added when non-selected item is clicked', () => {
      const selectedItemBtn = bubbleSelector.find('.notSelected').first();
      selectedItemBtn.simulate('click');
      expect(updateActiveSelection).toHaveBeenCalled();
      expect(updateActiveSelection).toHaveBeenCalledWith(['a', 'b', 'c'], 'add', 'c');
      updateActiveSelection.mockClear();
    });
  });
  describe('Adding custom entry', () => {
    const updateActiveSelection = jest.fn();
    const bubbleSelector = mount(
      <BubbleSelector
        activeSelection={['a', 'b', 'e']}
        options={[
          { uid: 'a', label: 'a' },
          { uid: 'b', label: 'b' },
          { uid: 'c', label: 'c' },
          { uid: 'd', label: 'd' },
        ]}
        updateActiveSelection={updateActiveSelection}
      />
    );
    it('allows selections outside of option set', () => {
      expect(bubbleSelector.find('.selected').length).toEqual(3);
    });
    // it('adds a selection on entering into text field and pressing enter', () => {
    //   // const textField =
    // });
  });
});
