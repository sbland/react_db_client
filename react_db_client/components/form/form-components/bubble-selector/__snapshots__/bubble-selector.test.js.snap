// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BubbleSelector Matches Snapshot - group selected 1`] = `
"<div className=\\"bubbleSelector\\">
  <ul className=\\"bubbleSelector_list selected\\">
    <BubbleSelectorItem uid=\\"a\\" label=\\"a\\" handleSelect={[Function: handleSelect]} isSelected={true} />
    <BubbleSelectorItem uid=\\"c\\" label=\\"c\\" handleSelect={[Function: handleSelect]} isSelected={true} />
  </ul>
  <ul className=\\"bubbleSelector_list unselected\\">
    <BubbleSelectorItem uid=\\"b\\" label=\\"b\\" handleSelect={[Function: handleSelect]} isSelected={false} />
    <BubbleSelectorItem uid=\\"d\\" label=\\"d\\" handleSelect={[Function: handleSelect]} isSelected={false} />
  </ul>
</div>"
`;

exports[`BubbleSelector Matches Snapshot 1`] = `
"<div className=\\"bubbleSelector\\">
  <ul className=\\"bubbleSelector_list notGrouped\\">
    <BubbleSelectorItem uid=\\"a\\" label=\\"a\\" handleSelect={[Function: handleSelect]} isSelected={true} />
    <BubbleSelectorItem uid=\\"b\\" label=\\"b\\" handleSelect={[Function: handleSelect]} isSelected={false} />
    <BubbleSelectorItem uid=\\"c\\" label=\\"c\\" handleSelect={[Function: handleSelect]} isSelected={true} />
    <BubbleSelectorItem uid=\\"d\\" label=\\"d\\" handleSelect={[Function: handleSelect]} isSelected={false} />
  </ul>
</div>"
`;
